### PR TITLE
In ES3 emit, use index notation for member access if the identifier is a reserved word

### PR DESCRIPTION
**Pull request by billti**
Per issue #2296, in ES3 reserved words may not be used as identifiers, even in member access positions.  This is a first attempt at a solution for this.

Looks like a number of baselines emit targeting ES3 and access members using reserved words, so a few are changed by this.  I haven't evaluated for perf impact yet.

I'll investigate further tomorrow.  Just wanted to get this out for feedback on approach, as I'm not overly familiar with the code and there may be more changes required or a cleaner solution.

### CODE REVIEW

File: src/compiler/emitter.ts [hunk 1]
```diff
@@ -55,6 +55,20 @@ module ts {
     }
 
     let indentStrings: string[] = ["", "    "];
+
+    // When emitting for ES3, property access via dot notation (i.e. obj.prop ) cannot be used
+    // if the property name is a reserved word.  Use index notation (i.e. obj["prop"]).
+    let ES3ReservedWords = [
```
Comment (CyrusNajmabadi) [inline src/compiler/emitter.ts]: I'd prefer us to just break up the SyntaxKind range to have the reserved words in a single range.  Then you can do this with just a SyntaxKind check.

File: src/compiler/emitter.ts [hunk 2]
```diff
@@ -3124,12 +3138,20 @@ module ts {
                     return;
                 }
 
+                var emitAsIndexed = (languageVersion === ScriptTarget.ES3 && IdentifierIsES3ReservedWord(node.name))
```
Comment (CyrusNajmabadi) [inline src/compiler/emitter.ts]: outer parens not necessary.

File: src/compiler/emitter.ts (hunk 1 above)
Comment (billti) [inline src/compiler/emitter.ts]: SyntaxKind is always Identifier, so this won't work (e.g. see line 69 below)

Comment (billti) [inline src/compiler/emitter.ts]: After cross-checking the list, the following are all ES3 FutureReservedWords, but not currently in the SyntaxKind list, which means we'd need to add them to SyntaxKind (and the textToTokenMap), even though we'll never parse/create a node of these kind. That seems less clean than the current design. Thoughts?

ES3 reserved words we currently don't have SyntaxKind members for: "abstract, int, short, boolean, byte, long, char, final, native, synchronized, float, throws, goto, transient, volatile"

Note: These are not keywords or reserved in ES5+

**Comment by billti** [comment]
Actually, after discussing further this is a little more involved, due to the fact that ES3 has reserved words that are not reserved words in ES5 and later.  Thus, a declaration such as "var abstract = 42;" should fail to emit in ES3, but is valid in ES5+.  Will think further on this...

File: src/compiler/emitter.ts (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/compiler/emitter.ts]: Based on our previous conversation, it sounds like we will need syntax kinds here.  That way we can report errors properly if you do things like write "function transient" and then try to emit for ES3.

**Comment by mhegazy** [comment]
@billti are we ready to take this in?

**Comment by mhegazy** [comment]
@billti  any news?

**Comment by schungx** [comment]
Any idea when this will come out?

**Comment by mhegazy** [comment]
@schungx, looks like @billti is busy with other issues at the moment. if someone is willing to takeover this PR we would be happy to help.

**Comment by schungx** [comment]
I would if I could...  *sigh*

But anyway, I think this is something that shouldn't be difficult to resolve and is extremely useful.  Those of us still targeting ES3 all have a deep dark secret...  we need to *sniff* support IE6-8.

**Comment by mhegazy** [comment]
Please refresh and resubmit.

**Pull request closed without merging**